packages/cli-repl/src/cli-repl.spec.ts

```diff
@@ -1,9 +1,189 @@
+        expect(JSON.parse(content).enableTelemetry).to.be.false;
+      });
+
+      it('emits exit when asked to', async() => {
+        const onexit = once(cliRepl.bus, 'mongosh:exit');
+        input.write('.exit\n');
+        await onexit;
+        expect(exitCode).to.equal(0);
+      });
+    });
+
+    context('during startup', () => {
+      it('persists userId', async() => {
+        const userIds: string[] = [];
+        await promisify(rimraf)(tmpdir.path);
+        for (let i = 0; i < 2; i++) {
+          cliRepl = new CliRepl(cliReplOptions);
+          cliRepl.bus.on('mongosh:new-user', userId => userIds.push(userId));
+          cliRepl.bus.on('mongosh:update-user', userId => userIds.push(userId));
+          await cliRepl.start('', {});
+        }
+        expect(userIds).to.have.lengthOf(2);
+        expect([...new Set(userIds)]).to.have.lengthOf(1);
+      });
+
+      it('emits error for invalid config', async() => {
+        await fs.writeFile(path.join(tmpdir.path, 'config'), 'notjson');
+        cliRepl = new CliRepl(cliReplOptions);
+        const onerror = once(cliRepl.bus, 'mongosh:error');
+        try {
+          await cliRepl.start('', {});
+        } catch { /* not empty */ }
+        await onerror;
+      });
+
+      it('emits error for inaccessible home directory', async() => {
+        cliReplOptions.shellHomePath = '/nonexistent/inaccesible';
+        cliRepl = new CliRepl(cliReplOptions);
+        const onerror = once(cliRepl.bus, 'mongosh:error');
+        try {
+          await cliRepl.start('', {});
+        } catch { /* not empty */ }
+        await onerror;
+      });
+
+      it('verifies the Node.js version', async() => {
+        delete (process as any).version;
+        process.version = 'v8.0.0';
+        try {
+          cliRepl = new CliRepl(cliReplOptions);
+          const onerror = once(cliRepl.bus, 'mongosh:error');
+          try {
+            await cliRepl.start('', {});
+          } catch { /* not empty */ }
+          await onerror;
+        } finally {
+          process.version = process.versions.node;
+        }
+      });
+    });
+  });
+
+  context('with an actual server', () => {
+    const testServer = startTestServer('shared');
+
+    it('connects to a server and interacts with it', async() => {
+      cliRepl = new CliRepl(cliReplOptions);
+      await cliRepl.start(await testServer.connectionString(), {});
+
+      output = '';
+      input.write('use clirepltest\n');
+      await waitEval(cliRepl.bus);
+      expect(output).to.include('switched to db clirepltest');
+
+      output = '';
+      input.write('db.cats.insertOne({name:"pia"})\n');
+      await waitEval(cliRepl.bus);
+      expect(output).to.include('acknowledged: true');
+
+      output = '';
+      input.write('db.cats.find()\n');
+      await waitEval(cliRepl.bus);
+      expect(output).to.include('pia');
+
+      input.write('.exit\n');
+    });
```

Comment:
This can do basically anything our e2e tests can also do, but without the overhead of spawning a new process or spinning a timer to wait for output from the child process. For some e2e tests, we could consider migrating them over to a pattern like this, if we want to.